In [3]:
# import csv
# with open('trip_data_7.csv') as f: 
#     data = [{k:str(v) for k, v in row.items()}
#            for row in csv.DictReader(f, skipinitialspace=True)]

In [7]:
import pandas as pd
data = pd.read_csv('trip_data_7.csv')
data.sample(10)

E:\apps\anaconda\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


medallion                      hack_license  \
1456979   59A3962EF65E5B44C804E842F5F7D04C  1D4EF564697EFE2C4CB60668988CDD23   
2245313   F3F85F66A0CB41DBF3386621C35F7030  86C610305A36FF53508F34C08D021DA4   
10784913  5E58DD84EB159B99D70618A36987F3AE  FCB409DEF4B657130794B94E6281F469   
11928285  539F70E88324AD888C52D6D58FF31AD6  2CE60CFF2F8A4CE0838D523C78D95934   
11846935  82A1032C25A2E4CFE244FF9976FBD3C9  77876603CB505E30C3F0E7BC9CDAB743   
5625271   A9DF8BDBEA3E57580A539596AA5D8B69  AD3A4A8ABECCCBA549D917BB47CCECE4   
5894872   0076C8327A95E988E721AC33B0FA9D67  7BDFF06419C23F667C5D69EEFBF091BE   
528983    6652CE5ACEF5E414C6FFEA71A8DB86F3  8333EAE5BE5879C45DB2CB0A2C7CF5F1   
4835906   8A9B39E0936ABB3D5F7933D187FC6655  DA964D248F82D1100B754D49D2A835B6   
1564421   F3B0E04AF07F5667AD64771DA60BD2FD  1CF5209424E06360D8ACD4C78730FE34   

          vendor_id   rate_code  store_and_fwd_flag      pickup_datetime  \
1456979         VTS           1                 NaN  2013-07-16 23:25:00   
2245313         VTS           1                 NaN  2013-07-03 22:39:00   
10784913        CMT           1                   N  2013-07-23 06:59:03   
11928285        CMT           1                   N  2013-07-30 16:12:45   
11846935        CMT           1                   N  2013-07-30 00:30:49   
5625271         VTS           1                 NaN  2013-07-27 04:16:00   
5894872         CMT           1                   N  2013-07-01 00:57:42   
528983          VTS           1                 NaN  2013-07-13 01:03:00   
4835906         VTS           1                 NaN  2013-07-23 23:18:00   
1564421         VTS           1                 NaN  2013-07-17 12:20:00   

             dropoff_datetime   passenger_count   trip_time_in_secs  \
1456979   2013-07-16 23:47:00                 1                1320   
2245313   2013-07-03 23:07:00                 1                1680   
10784913  2013-07-23 07:00:58                 1                 115   
11928285  2013-07-30 16:33:58                 1                1273   
11846935  2013-07-30 00:37:54                 1                 425   
5625271   2013-07-27 04:27:00                 5                 660   
5894872   2013-07-01 01:03:06                 1                 324   
528983    2013-07-13 01:11:00                 1                 480   
4835906   2013-07-23 23:29:00                 1                 660   
1564421   2013-07-17 12:50:00                 6                1800   

           trip_distance   pickup_longitude   pickup_latitude  \
1456979             4.92         -73.987663         40.748253   
2245313             4.17         -74.003624         40.723835   
10784913            0.50         -73.939659         40.810116   
11928285            3.40         -73.988518         40.754047   
11846935            1.50         -73.986732         40.729572   
5625271             3.64         -73.985245         40.758533   
5894872             1.90         -74.002159         40.740318   
528983              1.95         -74.012405         40.706726   
4835906             3.39         -73.972725         40.760452   
1564421             5.59         -73.980019         40.755356   

           dropoff_longitude   dropoff_latitude  
1456979           -73.951500          40.710869  
2245313           -73.980698          40.762962  
10784913          -73.938477          40.804726  
11928285          -73.956482          40.775444  
11846935          -74.005615          40.725803  
5625271           -73.947746          40.779224  
5894872           -73.985100          40.763668  
528983            -73.994377          40.717316  
4835906           -73.962524          40.797302  
1564421           -73.927177          40.770683

In [14]:
len(data)

13823840

In [31]:
df = data

In [32]:
# data[data.columns[0]].count()
df.columns.values

array(['medallion', ' hack_license', ' vendor_id', ' rate_code',
       ' store_and_fwd_flag', ' pickup_datetime', ' dropoff_datetime',
       ' passenger_count', ' trip_time_in_secs', ' trip_distance',
       ' pickup_longitude', ' pickup_latitude', ' dropoff_longitude',
       ' dropoff_latitude'], dtype=object)

In [36]:
# # 1. Create a list of columns to be selected
# columns_to_be_selected = [' pickup_datetime', ' dropoff_datetime']
# # 2. Use it as an index to the DataFrame
# df[columns_to_be_selected]
# # 3. Using loc method
# df.loc[columns_to_be_selected]

In [20]:
df[' pickup_datetime']

0           2013-07-01 01:47:00
1           2013-07-01 01:39:00
2           2013-07-01 01:36:00
3           2013-07-01 01:53:00
4           2013-07-01 01:47:00
                   ...         
13823835    2013-07-01 01:45:00
13823836    2013-07-01 01:17:00
13823837    2013-07-01 01:36:00
13823838    2013-07-01 01:46:00
13823839    2013-07-01 01:34:00
Name:  pickup_datetime, Length: 13823840, dtype: object

In [24]:
# formate is %Y-%m-%d %H:%M:%S
import datetime as dt
df[' pickup_datetime'] = pd.to_datetime(df[' pickup_datetime']) 
date_range = str(df[' pickup_datetime'].dt.date.min()) + ' to ' +str(df[' pickup_datetime'].dt.date.max())

In [30]:
print(date_range)
# df[' pickup_datetime'].dt.date   

2013-07-01 to 2013-07-31


In [38]:
# df.iloc[0:3, :]
df.head()

medallion                      hack_license  \
0  9406D2C34715E1DA10AD4D4DDADF4DA5  0602DFD837433635FE860BDE2F14BC3A   
1  73495B48481E673AD8D7578764C9EC98  66DA261DA2E0305435786C152085CCC2   
2  85B6168FC88F4914E4741015B1678BF4  C2DB36C1124AB69269DBE1D3D39D37DB   
3  85B6168FC88F4914E4741015B1678BF4  C2DB36C1124AB69269DBE1D3D39D37DB   
4  8C6CFF1ED34957DCF1547C3C38EE0490  65AB951A1B846C3B0581A8A473830F73   

   vendor_id   rate_code  store_and_fwd_flag     pickup_datetime  \
0        VTS           1                 NaN 2013-07-01 01:47:00   
1        VTS           1                 NaN 2013-07-01 01:39:00   
2        VTS           1                 NaN 2013-07-01 01:36:00   
3        VTS           1                 NaN 2013-07-01 01:53:00   
4        VTS           1                 NaN 2013-07-01 01:47:00   

      dropoff_datetime   passenger_count   trip_time_in_secs   trip_distance  \
0  2013-07-01 01:52:00                 1                 300            1.17   
1  2013-07-01 01:52:00                 1                 780            4.17   
2  2013-07-01 01:50:00                 1                 840            2.98   
3  2013-07-01 01:57:00                 1                 240            0.75   
4  2013-07-01 01:59:00                 1                 720            1.76   

    pickup_longitude   pickup_latitude   dropoff_longitude   dropoff_latitude  
0         -74.013229         40.714718          -74.003494          40.725285  
1         -73.980865         40.763947          -73.933884          40.767677  
2         -74.007568         40.740898          -73.980118          40.713985  
3         -73.985466         40.718342          -73.982841          40.727016  
4         -74.005630         40.726448          -74.003906          40.742031

In [43]:
df.describe()

rate_code   passenger_count   trip_time_in_secs   trip_distance  \
count  1.382384e+07      1.382384e+07        1.382384e+07    1.382384e+07   
mean   1.032433e+00      1.724331e+00        7.495074e+02    2.904796e+00   
std    3.125351e-01      1.388857e+00        5.565799e+02    3.381533e+00   
min    0.000000e+00      0.000000e+00        0.000000e+00    0.000000e+00   
25%    1.000000e+00      1.000000e+00        3.600000e+02    1.070000e+00   
50%    1.000000e+00      1.000000e+00        6.000000e+02    1.800000e+00   
75%    1.000000e+00      2.000000e+00        9.600000e+02    3.200000e+00   
max    2.100000e+02      8.000000e+00        1.080000e+04    1.000000e+02   

        pickup_longitude   pickup_latitude   dropoff_longitude  \
count       1.382384e+07      1.382384e+07        1.382314e+07   
mean       -7.288411e+01      4.014884e+01       -7.283900e+01   
std         8.999069e+00      5.555792e+00        9.190940e+00   
min        -2.355738e+03     -3.114562e+03       -2.352800e+03   
25%        -7.399251e+01      4.073457e+01       -7.399177e+01   
50%        -7.398219e+01      4.075203e+01       -7.398090e+01   
75%        -7.396814e+01      4.076572e+01       -7.396532e+01   
max         1.200160e+02      3.210367e+03        1.428736e+03   

        dropoff_latitude  
count       1.382314e+07  
mean        4.012345e+01  
std         5.515097e+00  
min        -3.447916e+03  
25%         4.073373e+01  
50%         4.075246e+01  
75%         4.076633e+01  
max         4.739997e+02

In [61]:
df[" vendor_id"].unique()

array(['VTS', 'CMT'], dtype=object)

In [56]:
df[" rate_code"].unique()

array([  1,   5,   3,   2,   4,   8,   0,   6, 210,   7,   9], dtype=int64)

In [57]:
df[" store_and_fwd_flag"].unique()

array([nan, 'N', 'Y'], dtype=object)

In [59]:
df[" passenger_count"].unique()

array([1, 5, 4, 6, 2, 3, 0, 8], dtype=int64)

In [65]:
print('rate_code min:' , df[" rate_code"].min())
print('rate_code max:' , df[" rate_code"].max())

rate_code min: 0
rate_code max: 210


In [ ]:
print('rate_code min:' , df[" rate_code"].min())
print('rate_code max:' , df[" rate_code"].max())

In [67]:
print('trip_distance min:' , df[" trip_distance"].min())
print('trip_distance max:' , df[" trip_distance"].max())

trip_distance min: 0.0
trip_distance max: 100.0


In [55]:
df[" trip_distance"].max()

100.0

| medallion| hack_license| vendor_id| rate_code| store_and_fwd_flag| pickup_datetime| dropoff_datetime| passenger_count| trip_time_in_secs| trip_distance| pickup_longitude| pickup_latitude| dropoff_longitude| dropoff_latitude|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 9406D2C34715E1DA10AD4D4DDADF4DA5 | 0602DFD837433635FE860BDE2F14BC3A | VTS | 1 | NaN | 2013-07-01 01:47:00 | 2013-07-01 01:52:00 | 1 | 300 | 1.17 | -74.013229 | 40.714718 | -74.003494 | 40.725285 |
|73495B48481E673AD8D7578764C9EC98 | 66DA261DA2E0305435786C152085CCC2 | VTS | 1 | NaN | 2013-07-01 01:39:00 | 2013-07-01 01:52:00 | 1 | 780 | 4.17 | -73.980865 | 40.763947 | -73.933884 | 40.767677|
|85B6168FC88F4914E4741015B1678BF4 | C2DB36C1124AB69269DBE1D3D39D37DB | VTS | 1 | NaN | 2013-07-01 01:36:00 | 2013-07-01 01:50:00 | 1 | 840 | 2.98 | -74.007568 | 40.740898 | -73.980118 | 40.713985|


                

In [44]:
dict(zip(df.columns,df.dtypes))

{'medallion': dtype('O'),
 ' hack_license': dtype('O'),
 ' vendor_id': dtype('O'),
 ' rate_code': dtype('int64'),
 ' store_and_fwd_flag': dtype('O'),
 ' pickup_datetime': dtype('<M8[ns]'),
 ' dropoff_datetime': dtype('O'),
 ' passenger_count': dtype('int64'),
 ' trip_time_in_secs': dtype('int64'),
 ' trip_distance': dtype('float64'),
 ' pickup_longitude': dtype('float64'),
 ' pickup_latitude': dtype('float64'),
 ' dropoff_longitude': dtype('float64'),
 ' dropoff_latitude': dtype('float64')}

| Field | dtype (SQL)    |
| :---:   | :---: |
| medallion | varchar(50)  |
| hack_license | varchar(50) |
| vendor_id | varchar(5) |
| rate_code | int(10) |
| store_and_fwd_flag | bool |
| pickup_datetime | datetime |
| dropoff_datetime | datetime |
| passenger_count | int(10) |
| trip_time_in_secs | int(10) |
| trip_distance | Decimal(5,5) |
| pickup_longitude | Decimal(9,6) |
| pickup_latitude | Decimal(8,6) |
| dropoff_longitude | Decimal(9,6) |
| dropoff_latitude | Decimal(8,6) |
